### ISSUE TITLE
`@scalar/fastify-api-reference` doesn't seem to work when bundling

### ISSUE THREAD

**Issue by acidoxee** [scalar/scalar#3566]
### What happens?

Hi, we're currently using self-hosted Scalar with the React integration for OpenAPI docs, and wanted to replace this with serving the docs directly from our Fastify server, as we'll need to document several API versions, and each one being served directly from Fastify would be easier and doesn't require setting up an app.

We've tried using `@scalar/fastify-api-reference`, and this works great locally. But for production, we're bundling our code and running it from a Docker container, and it seems to not work in this state, with the following error:

![scalar js issue with bundling](https://github.com/user-attachments/assets/11e7caa8-95fd-4815-9f1c-cf875c9fd33c)

### What did you expect to happen?

We expected it to work just like locally, but it doesn't. I'm guessing this "standalone" JS file is expected to be served with the Fastify API or available in the filesystem, but this isn't the case when bundling.

Note: This is a duplicate of the [Discord thread](https://discord.com/channels/1135330207960678410/1263945110052274187/threads/1296449225369976903) I opened a few days ago.

**Comment by hanspagel** [comment]
We’re using `import.meta.url` which is only available in ESM environments, maybe we need to add a CommonJS polyfill.

**Comment by acidoxee** [comment]
Absolutely! It's extremely minimal though, as it doesn't handle the build part (we have a separate CI step that handles it along with the build of our other commands), and so only copies the bundled assets (or should I say, _asset_, since there is only a single file left after bundling):

```Dockerfile
FROM node:20.18.0-alpine

WORKDIR /usr/src/app
COPY . .

EXPOSE 80

CMD ["node", "main.mjs"]
```

I also confirm we're ESM all the way. Our source code is entirely ESM-only, and for external modules we do favor ESM entrypoints first when bundling with esbuild (with `mainFields: ['exports', 'module', 'main']`, as opposed to the default behavior of using a CJS entrypoint if there is one).

**Comment by hanspagel** [comment]
Okay, that’s probably the problem. We’re reading the file from the node_modules folder, but if you just have one file it’s just not there anymore.

**Comment by anat** [comment]
Same, not bundlable.

Can run using tsx but not using ts-node.

**Comment by zachstence** [comment]
I ran into this issue when trying to run a bun executable

```
/$bunfs/root/js/standalone.js
error: JavaScript file not found: /$bunfs/root/js/standalone.js
      at getJavaScriptFile (/$bunfs/root/exec:67549:11)
      at <anonymous> (/$bunfs/root/exec:67788:42)
      at <anonymous> (/$bunfs/root/exec:67751:64)
      at <anonymous> (/$bunfs/root/exec:30082:34)
      at <anonymous> (/$bunfs/root/exec:30575:16)
      at <anonymous> (/$bunfs/root/exec:30555:25)

Bun v1.2.5 (Linux x64)
```

**Comment by tjmcdonough** [comment]
Same for me too. moduleResolution doesn't work with NodeNext, just node

**Comment by psyirius** [comment]
I made a workaround using pnpm patch
```diff
--- a/dist/utils/getJavaScriptFile.js
+++ b/dist/utils/getJavaScriptFile.js
@@ -1,9 +1,12 @@
 import fs from "node:fs";
 import path from "node:path";
 import { fileURLToPath } from "node:url";
+// making it bundleable by vite
+import jsFile from '../js/standalone.js?url&no-inline'
 function getJavaScriptFile() {
   const dirname = path.dirname(fileURLToPath(import.meta.url));
   const filePath = [
+    path.join(dirname, path.basename(jsFile)), // for vite, it will be under /chunks where this file (bundled) also resides
     path.resolve(`${dirname}/js/standalone.js`),
     path.resolve(`${dirname}/../../dist/js/standalone.js`)
   ].find((file) => fs.existsSync(file));
```

This works for my vite setup (where the referencing bundle.js (worker) also resides in the same /chunks dir)
```js
import jsFile from '../js/standalone.js?url&no-inline'
/// jsFile  => '/chunks/standalone-<HASH>.js'
```

Caveats:
- Works only when processed through vite

**Comment by hanspagel** [comment]
I just tried to get it working for 2 hours, but failed. It seems so easy, this is what we want:

* don’t have @scalar/api-reference as a dependency
* add the bundled JS from @scalar/api-reference to the fastify package
* inline it as a string, so even if your bundler does tree-shaking and everything, it’s still there
* we don’t want to put it in the src/ folder, otherwise the browser build has to go through the esbuild pipeline (which is configured for Node)

I tried esbuild settings and plugins, wrote custom plugins, nothing worked (yet).

**Comment by psyirius** [comment]
> I just tried to get it working for 2 hours, but failed. It seems so easy, this is what we want:
> 
> * don’t have @scalar/api-reference as a dependency
> * add the bundled JS from @scalar/api-reference to the fastify package
> * inline it as a string, so even if your bundler does tree-shaking and everything, it’s still there
> * we don’t want to put it in the src/ folder, otherwise the browser build has to go through the esbuild pipeline (which is configured for Node)
> 
> I tried esbuild settings and plugins, wrote custom plugins, nothing worked (yet).

Did you tried patching the `utils/getJavaScriptFile.js` file?

You can use:
- A pnpm patch like I mentioned above.
- A rollup/vite plugin to patch the source content of the 'utils/getJavaScriptFile.js' when it's getting resolved.
